src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -4302,6 +4302,48 @@ void C2_MacroAssembler::vector_mask_operation(int opc, Register dst, XMMRegister
 }
 #endif
 
+void C2_MacroAssembler::vector_signum_evex(int opcode, XMMRegister dst, XMMRegister src, XMMRegister zero, XMMRegister one,
+                                           XMMRegister xtmp1, KRegister ktmp1, KRegister ktmp2, int vec_enc) {
+  if (opcode == Op_SignumVD) {
+    evfpclasspd(ktmp1, src, 0x7, vec_enc);
+    evpmovq2m(ktmp2, src, vec_enc);
+    evsubpd(xtmp1, k0, zero, one, true, vec_enc);
+    evmovdquq(dst, k0, one, true, vec_enc);
+    evmovdquq(dst, ktmp2, xtmp1, true, vec_enc);
+    evblendmpd(dst, ktmp1, dst, src, true, vec_enc);
```

Comment:
I believe you could achieve better register management using the following sequence. Also I would suggest using `evmovupd`or `evmovups` to avoid potential data bypass delay.

    evpmovq2m(ktmp1, src, vec_enc);
    evmovdquq(dst, k0, one, false, vec_enc);
    evsubpd(dst, ktmp1, zero, one, true, vec_enc);
    evfpclasspd(ktmp1, src, 0x7, vec_enc);
    evblendmpd(dst, ktmp1, dst, zero, true, vec_enc);

Reply:
> I believe you could achieve better register management using the following sequence.
> 
> ```
> evfpclasspd(ktmp1, src, 0x7, vec_enc);
> evblendmpd(dst, ktmp1, dst, zero, true, vec_enc);

0x7 encodes QNaN, +/- 0.0 values. Thus blending dst with zero will not work for NaN. I guess you wanted to mention src as in original sequence.

> evfpclasspd(ktmp1, src, 0x40, vec_enc);

0x40 checks does not check for NEGATIVE_INFINITE, Math.signum should return -1 for it.  0x40 should be 0x50

> evsubpd(dst, ktmp1, zero, one, true, vec_enc);
> ```

But I agree we can do away with some temporaries.

Reply:
Fixed the suggestion because I was mistaken, the first blend put all `NaN` and nonnegative values and the masked `vsubpd` put the remaining negative values, completing the operation. You need to spot `SNaN` also because it can be produced using `Double.longBitsToDouble` and in the future using vector reinterpretation.
Thanks.